### ISSUE TITLE
"SecurityError: The operation is insecure." from LocalStorage reading

### ISSUE THREAD

**Issue by Mikolaj** [gamejolt/issue-tracker#1076]
When I run HTML5 version of this game in Firefox 57.0 on Ubuntu (in Chrome it works fine)

https://gamejolt.com/games/allureofthestars/305565

the moment it tries to `getItem` from LocalStorage at this point in the code

https://github.com/LambdaHack/LambdaHack/blob/v0.7.1.0/Game/LambdaHack/Common/JSFile.hs#L53

it crashes with `SecurityError: The operation is insecure.`

The same code works fine (both on Firefox and Chrome) at http://allureofthestars.com/play

I vaguely suspect this is caused by https://en.wikipedia.org/wiki/Same-origin_policy
and the fact, that you put my JS into an iframe and then I try to access the LocalStorage from within the iframe. How to avoid this problem? Thank you for your assistance.

**Comment by YLivay** [comment]
Hey sorry for the slow response. The game is run from within an iframe under *.gamejolt.net. This is for security reasons so that the games won't have access to the user's local storage, cookie and other session info for their *.gamejolt.com domain. The specific domain for browser games is b-cdn.gamejolt.net and works fully with https.

I'm not sure how you're getting the localstorage reference, but if it doesn't use the correct origin it would fail. How is your game set up? Does it request resources from your site instead of the bundled resources in the game archive? Are you iframing it further?

**Comment by YLivay** [comment]
By the way, game looks really fun - i love creative text based games haha.
How can I trigger the security error?

**Comment by Mikolaj** [comment]
Thank you. :) I get the error every time I run the game from your site on Firefox at the address I gave (but you need to be patient, it takes a while to load, it's 7MB of JS and it seem to be not (always) cached coming from your site).

I'm not iframing the game further nor trying to access anything from my site.

When I run the same under Chrome, it indeed puts all savegames into `b-cdn.gamejolt.net`, so it would seem I try to access to correct local storage, at least under Chrome. Not sure how it may be related, but on Chrome nothing is written to gamejolt.com local storage, while on Firefox, there are some entries and I'm sure they don't come from me (chat-room-something). Anyway, the crash certainly occurs in my code, at the place I linked to.

Here is the JS foreign call code that obtains the "Window" that I subsequently use to access the local storage:
https://github.com/ghcjs/ghcjs-dom/blob/d13e77f6f8252db8c5e3e1502f90fe5d3902fa87/ghcjs-dom-jsffi/src/GHCJS/DOM.hs#L29

I've just replaced my game with a simpler one, where I obtain the current window directly via `getDefaultView` applied to `currentDocument` and where the JS code is not obfuscated by the Closure Compiler (ccjs), but the crash is the same.

**Comment by Mikolaj** [comment]
O, dear. I changed "Accept cookies from websites" from "From Visited" to "Always" and now it works. I think "From Visited" is the default nowadays, so people will have that problem.

What would you suggest in that case? Apparently one of the sites that are involved in the loading and running the game is not "Visited". Can it be made "Visited"? Can I change the code in any way? Could we warn Firefox users to change their cookies setting?

**Comment by hworld** [comment]
Hmm, so Firefox's default is to allow 3rd party cookies since many web sites require them to work to allow logging in. Here's the article they have for disabling: https://support.mozilla.org/en-US/kb/disable-third-party-cookies

I just checked on a fresh install of Firefox and it allows 3rd party.

From what I know, the only browser that doesn't allow them by default is Safari. We actually have a work-around for Safari in the code here: https://github.com/gamejolt/frontend-lib/blob/45b252d9e7ca6dfa3041b0b0fb848b9fed38ceab/components/game/play-modal/play-modal.service.ts#L56

It basically pops it out in a new window instead of showing in the iframe. The better way to handle this maybe is to try to set/read a 3rd party cookie when trying to show the game, and if it doesn't work, then we pop it out in a new window and do that separate flow.

**Comment by Mikolaj** [comment]
Yes, I had 3rd party cookies enabled, but only for Visited sites. Various help pages suggest this setting should be enough and it's probably the default (what setting did you see in your fresh install?). I had to change it be Always enabled and only then it works. Thank you for looking at this!

**Comment by YLivay** [comment]
Very interesting about the caching by the way. Each build has its own hash in the url (build folder hash) so we should be able to have our storage return cache headers for a long time. I'll look into it too :)

**Comment by YLivay** [comment]
These are the default settings im getting on a FF clean installation.
![image](https://user-images.githubusercontent.com/996947/34511987-e8c7dccc-f067-11e7-9b3e-4b08e039f7b1.png)

Perhaps it imported a setting from the previous browser when it was installed? Chrome, Edge and IE9-11 allow third party cookies by default too.

At any rate I think @hworld solution would be the best solution so we wouldn't have to worry about browsers changing their minds about the matter haha..

**Comment by Mikolaj** [comment]
Oh, so it's Always by default, not Visited. Good. Thank you again. So I must have changed the cookies based on some alarming security article and then I forgot. If the game works fine with the default settings, then I won't refrain from publishing it. Any workaround, though, or a hint for the user to switch from Visited to Always would be obviously welcome. I may not be the only over-cautious user. :)

Regarding caching of the JS, you are right, it caches fine, I tested both with the unobfuscated 24M JS and the 7M JS. The delay must be the slow Firefox JS engine loading and setting up the JS --- I was used to Chrome that does it much faster.

Thank you guys, you rock. My confidence in Gamejolt soared. :)

**Comment by YLivay** [comment]
Thank you too for the patience :)

I'm leaving the issue open because we should definitely pop the game out in its own tab if we can detect the cookies are not accessible. Renamed the title of the issue to reflect that.
